decode.go

```diff
@@ -299,6 +299,28 @@ func (p *parser) mapping() *Node {
 // ----------------------------------------------------------------------------
 // Decoder, unmarshals a node into a provided value.
 
+// DecodeConfig represents configuration for decoding.
+// Config could be initialized in "builder" style for convenience:
+// NewDecodeConfig().KnownFields(true).UniqueFields(true)
```

Comment:
One of the reasons why I've used the method style in the encoder and decoder is that it allows us to protect against modification of the configuration out of turn. It also means we can use an initializer such as NewEncoder/NewDecoder, which prevents us having to test for initialization everywhere else.

That said, the Options type will by definition be a type prepared ahead of the creation. It seems like it would benefit from using a structure with public fields instead. It should have just one private field to force a non-positional instantiation.

Reply:
Not fully sure that I follow this. I added `DecodeOptions` argument to the internal `newDecoder` function. Is that what you meant above?